notebooks/ch11/gradient_stochastic.ipynb

```diff
@@ -0,0 +1,285 @@
+    }
+   },
+   "outputs": [],
+   "source": [
+    "# HIDDEN\n",
+    "tips = sns.load_dataset('tips')\n",
+    "tips['pcttip'] = tips['tip'] / tips['total_bill'] * 100"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 3,
+   "metadata": {
+    "ExecuteTime": {
+     "end_time": "2018-06-26T05:26:51.269266Z",
+     "start_time": "2018-06-26T05:26:51.266035Z"
+    }
+   },
+   "outputs": [],
+   "source": [
+    "# HIDDEN\n",
+    "def mse_cost(theta, y_vals):\n",
+    "    return np.mean((y_vals - theta) ** 2)\n",
+    "\n",
+    "def grad_mse_cost(theta, y_vals):\n",
+    "    return -2 * np.mean(y_vals - theta)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## A Brief Review\n",
+    "\n",
+    "We initially modeled the tips dataset by heuristically calculating the $\\hat\\theta$ that minimized the MSE cost function, $L(\\hat\\theta, y) = \\frac{1}{n} \\sum_{i=1}^{n}(y_i - \\hat\\theta)^2$. This involved taking the derivative of the MSE, setting it to zero, and solving for $\\hat\\theta$. We found that the MSE-minimizing $\\hat\\theta$ value was simply the mean of the $y$ values in our dataset.\n",
+    "\n",
+    "But for more complicated models and cost functions, there may not be simple algebraic expressions that yield the cost-minimizing $\\hat\\theta$ values. Our solution was a new method called gradient descent, which iteratively updates $\\hat\\theta$ until it doesn't change between iterations. To complete an iteration of gradient descent, we calculate the following:\n",
+    "\n",
+    "$$\n",
+    "\\hat\\theta_{t+1} = \\hat\\theta_t - \\alpha \\cdot \\nabla_{\\hat\\theta} L(\\hat\\theta, y)\n",
+    "$$"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Limitations of Gradient Descent\n",
+    "\n",
+    "Often, we calculate $\\nabla_{\\hat\\theta}L(\\hat\\theta, y)$ using the average gradient of the loss function $l(\\hat\\theta, y_i)$ across the entire dataset as a single batch. For this reason, this gradient update rule is often referred to as **batch gradient descent**.\n",
+    "\n",
+    "For example, the gradient of the MSE cost first requires us to find the gradient of the squared loss, $\\nabla_{\\hat\\theta} l(\\hat\\theta, y_i) = -2 (y_i - \\hat\\theta)$, for each of the $n$ observations in our dataset. We then compute the average of these results. This is shown below:\n",
+    "\n",
+    "$$\n",
+    "\\begin{align}\n",
+    "\\nabla_{\\hat\\theta} L(\\hat\\theta, y) &= -\\frac{2}{n} \\sum_{i=1}^{n}(y_i - \\hat\\theta) \\\\\n",
+    "&= \\frac{1}{n} \\sum_{i=1}^{n}-2(y_i - \\hat\\theta) \\\\\n",
+    "&= \\frac{1}{n} \\sum_{i=1}^{n} \\nabla_{\\hat\\theta} l(\\hat\\theta, y_i)\n",
+    "\\end{align}\n",
+    "$$\n",
+    "\n",
+    "\n",
+    "\n",
+    "This process is then repeated at each iteration $t$ of the gradient descent algorithm. For large $n$, this can become a computationally expensive problem to solve."
```

Comment:
Referring to $ t $ here is a little confusing since it isn't used in the above derivation, so perhaps this can be written without the $ t $: "... at each iteration of the gradient ..."

Instead when you mention the general equation involving theta_t+1 and theta_t in the above section, you could clarify that theta_t is the current choice/value of theta and theta_t+1 is the next choice/value of theta (also add a brief reminder with this that alpha is the learning rate).

Reply:
Clarified in the "Brief Review" section what each of the variables means; it should be clearer what $t$ is, so I've left $t$ in for "This process is then repeated at each iteration $t$ of the gradient descent algorithm"